lemur/certificates/service.py

```diff
@@ -1005,3 +1009,105 @@ def remove_destination_association(certificate, destination):
+    """
+    all_certs = defaultdict(dict)
+    for c in get_certs_for_expiring_deployed_cert_check(exclude):
+        domains_for_cert = find_and_persist_domains_where_cert_is_deployed(c, timeout_seconds_per_network_call)
+        if len(domains_for_cert) > 0:
+            all_certs[c] = domains_for_cert
+    database.commit()
+
+
+def get_certs_for_expiring_deployed_cert_check(exclude):
+    now = arrow.utcnow()
+    threshold_days = current_app.config.get("LEMUR_EXPIRING_DEPLOYED_CERT_THRESHOLD_DAYS", 14)
+    max_not_after = now + timedelta(days=threshold_days)
+
+    q = (
+        database.db.session.query(Certificate)
+        .filter(Certificate.not_after <= max_not_after)
+        .filter(Certificate.expired == false())
+        .filter(Certificate.revoked == false())
+        .filter(Certificate.in_rotation_window == true())
+    )
+
+    exclude_conditions = []
+    if exclude:
+        for e in exclude:
+            exclude_conditions.append(~Certificate.name.ilike("%{}%".format(e)))
+
+        q = q.filter(and_(*exclude_conditions))
+
+    return windowed_query(q, Certificate.id, 10000)
+
+
+def find_and_persist_domains_where_cert_is_deployed(certificate, timeout_seconds_per_network_call):
+    """
+    Checks if the specified cert is still deployed. Returns a list of domains to which it's deployed.
+
+    We use the serial number to identify that a certificate is identical. If there were multiple certificates
+    issued for the same domain with identical serial numbers, this could return a false positive.
+
+    Note that this checks *all* configured ports (specified in config LEMUR_PORTS_FOR_DEPLOYED_CERTIFICATE_CHECK)
+    for all the domains in the cert. If the domain is valid but the port is not, we have to wait for the connection
+    to time out, which means this can be quite slow.
+
+    :return: A dictionary of the form {'domain1': [ports], 'domain2': [ports]}
+    """
+    matched_domains = defaultdict(list)
+    # filter out wildcards, we can't check them
+    for cert_association in [assoc for assoc in certificate.certificate_associations if '*' not in assoc.domain.name]:
+        domain = cert_association.domain
+        matched_ports_for_domain = []
+        for port in current_app.config.get("LEMUR_PORTS_FOR_DEPLOYED_CERTIFICATE_CHECK", [443]):
+            try:
+                parsed_serial = parse_serial(get_certificate_via_tls(domain.name, port,
+                                                                     timeout_seconds_per_network_call))
+                if parsed_serial == int(certificate.serial):
+                    matched_ports_for_domain.append(port)
+            except Exception:
+                current_app.logger.info(f'Unable to check certificate for domain {domain} on port {port}',
+                                        exc_info=True)
+        if len(matched_ports_for_domain) > 0:
+            matched_domains[domain.name] = matched_ports_for_domain
+            # Update the DB
```

Comment:
would it help to have the DB update take place here?

Reply:
The DB update is happening here for the individual cert (line 1082 `cert_association.ports = matched_ports_for_domain` is the update), but I placed the commit outside after _all_ the certs are modified. I'm not actually sure if it's preferable to commit once or many times; if there's not reason to avoid many individual commits, then committing here would be "safer" in that there's less risk of a single cert preventing all the others from being updated.

Reply:
I just moved it to commit after every update, but let me know if you have other thoughts.

Reply:
yeah, +1 one committing more granularly, to avoid losing the entire state.